src/stress_stream/consumer.py

```diff
@@ -0,0 +1,82 @@
+import asyncio
+import struct
+from datetime import datetime
+
+import typer
+from prometheus_client import Counter, Histogram
+from streamstore import S2
+from streamstore.schemas import SequencedRecord
+
+from stress_stream._constants import HISTOGRAM_BUCKETS, METRICS_NAMESPACE, S2_AUTH_TOKEN
+from stress_stream._utils import metrics_server
+
+e2e_latency_histogram = Histogram(
+    name="e2e_latency_seconds",
+    documentation="End-to-end latency in seconds. This measures the time taken from when a record is produced "
+    "(by the producer) to when it is read by the consumer.",
+    namespace=METRICS_NAMESPACE,
+    buckets=HISTOGRAM_BUCKETS,
+)
+read_records_counter = Counter(
+    name="read_records_total",
+    documentation="Total number of records successfully read.",
+    namespace=METRICS_NAMESPACE,
+)
+read_bytes_counter = Counter(
+    name="read_bytes_total",
+    documentation="Total number of bytes successfully read.",
+    namespace=METRICS_NAMESPACE,
+)
+read_errors_counter = Counter(
+    name="read_errors_total",
+    documentation="Total number of failed read session operations.",
+    namespace=METRICS_NAMESPACE,
+)
+
+
+def observe_metrics(record: SequencedRecord):
+    name, val = record.headers[0]
+    read_bytes_counter.inc(len(record.body) + len(name) + len(val))
+    created_ts = struct.unpack("d", val)[0]
+    e2e_latency = (datetime.now() - datetime.fromtimestamp(created_ts)).total_seconds()
+    e2e_latency_histogram.observe(e2e_latency)
+
+
+async def consumer(basin_name: str, stream_name: str):
+    async with S2(auth_token=S2_AUTH_TOKEN) as s2:
+        stream = s2[basin_name][stream_name]
+        start_seq_num = await stream.check_tail()
+        try:
+            with read_errors_counter.count_exceptions():
+                async for output in stream.read_session(start_seq_num):
+                    match output:
+                        case list(records):
+                            read_records_counter.inc(len(records))
+                            for record in records:
+                                observe_metrics(record)
+                        case unexpected:
+                            raise RuntimeError(
+                                "Unexpected output when reading from stream. "
+                                f"Expected a batch of records, but got {unexpected}"
+                            )
+        finally:
+            # NB: The purpose of sleep here is to ensure that Prometheus gets a chance to scrape
+            # the read errors counter value.
```

Comment:
Ah instead of sleeping, I think just have a `while True` forever loop for retries with a blanket `except` handler to bump the error counter. Similar for producer.

Reply:
Oh, just saw https://github.com/s2-streamstore/stress-stream/pull/1#discussion_r1934924698

Re: producer-side, makes sense. Re: consumer, I do think we should try to avoid having to wait on sleep + restart to get going again...

Reply:
cool. will just do a `while True` then.